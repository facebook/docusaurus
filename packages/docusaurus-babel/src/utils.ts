/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import fs from 'fs-extra';
import path from 'path';
import {BABEL_CONFIG_FILE_NAME} from '@docusaurus/utils';
import type {TransformOptions} from '@babel/core';

export async function getCustomBabelConfigFilePath(
  siteDir: string,
): Promise<string | undefined> {
  const customBabelConfigurationPath = path.join(
    siteDir,
    BABEL_CONFIG_FILE_NAME,
  );
  return (await fs.pathExists(customBabelConfigurationPath))
    ? customBabelConfigurationPath
    : undefined;
}

export function getBabelOptions({
  isServer,
  babelOptions,
}: {
  isServer?: boolean;
  // TODO Docusaurus v4 fix this
  //  weird to have getBabelOptions take a babelOptions param
  babelOptions?: TransformOptions | string;
} = {}): TransformOptions {
  const caller = {name: isServer ? 'server' : 'client'};
  if (typeof babelOptions === 'string') {
    return {
      babelrc: false,
      configFile: babelOptions,
      caller,
    };
  }
  return {
    ...(babelOptions ?? {
      presets: [require.resolve('@docusaurus/babel/preset')],
    }),
    babelrc: false,
    configFile: false,
    caller,
  };
}
