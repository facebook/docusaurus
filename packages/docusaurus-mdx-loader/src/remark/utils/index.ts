/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import escapeHtml from 'escape-html';
import type {Parent, Node} from 'unist';
import type {PhrasingContent, Heading} from 'mdast';
import type {
  MdxJsxAttribute,
  MdxJsxAttributeValueExpression,
  MdxJsxTextElement,
  // @ts-expect-error: TODO see https://github.com/microsoft/TypeScript/issues/49721
} from 'mdast-util-mdx';

/**
 * Util to transform one node type to another node type
 * The input node is mutated in place
 * @param node the node to mutate
 * @param newNode what the original node should become become
 */
export function transformNode<NewNode extends Node>(
  node: Node,
  newNode: NewNode,
): NewNode {
  Object.keys(node).forEach((key) => {
    // @ts-expect-error: unsafe but ok
    delete node[key];
  });
  Object.keys(newNode).forEach((key) => {
    // @ts-expect-error: unsafe but ok
    node[key] = newNode[key];
  });
  return node as NewNode;
}

export function stringifyContent(
  node: Parent,
  toString: (param: unknown) => string, // TODO weird but works
): string {
  return (node.children as PhrasingContent[])
    .map((item) => toValue(item, toString))
    .join('');
}

// TODO This is really a workaround, and not super reliable
// For now we only support serializing tagName, className and content
// Can we implement the TOC with real JSX nodes instead of html strings later?
function mdxJsxTextElementToHtml(
  element: MdxJsxTextElement,
  toString: (param: unknown) => string, // TODO weird but works
): string {
  const tag = element.name;

  const attributes = element.attributes.filter(
    (child): child is MdxJsxAttribute => child.type === 'mdxJsxAttribute',
  );

  const classAttribute =
    attributes.find((attr) => attr.name === 'className') ??
    attributes.find((attr) => attr.name === 'class');

  const classAttributeString = classAttribute
    ? `class="${escapeHtml(String(classAttribute.value))}"`
    : ``;

  const allAttributes = classAttributeString ? ` ${classAttributeString}` : '';

  const content = stringifyContent(element, toString);

  return `<${tag}${allAttributes}>${content}</${tag}>`;
}

export function toValue(
  node: PhrasingContent | Heading | MdxJsxTextElement,
  toString: (param: unknown) => string, // TODO weird but works
): string {
  switch (node.type) {
    case 'mdxJsxTextElement': {
      return mdxJsxTextElementToHtml(node as MdxJsxTextElement, toString);
    }
    case 'text':
      return escapeHtml(node.value);
    case 'heading':
      return stringifyContent(node, toString);
    case 'inlineCode':
      return `<code>${escapeHtml(node.value)}</code>`;
    case 'emphasis':
      return `<em>${stringifyContent(node, toString)}</em>`;
    case 'strong':
      return `<strong>${stringifyContent(node, toString)}</strong>`;
    case 'delete':
      return `<del>${stringifyContent(node, toString)}</del>`;
    case 'link':
      return stringifyContent(node, toString);
    default:
      return toString(node);
  }
}

export function assetRequireAttributeValue(
  requireString: string,
  hash: string,
): MdxJsxAttributeValueExpression {
  return {
    type: 'mdxJsxAttributeValueExpression',
    value: `require("${requireString}").default${hash && ` + '${hash}'`}`,
    data: {
      estree: {
        type: 'Program',
        body: [
          {
            type: 'ExpressionStatement',
            expression: {
              type: 'BinaryExpression',
              left: {
                type: 'MemberExpression',
                object: {
                  type: 'CallExpression',
                  callee: {
                    type: 'Identifier',
                    name: 'require',
                  },
                  arguments: [
                    {
                      type: 'Literal',
                      value: requireString,
                      raw: `"${requireString}"`,
                    },
                  ],
                  optional: false,
                },
                property: {
                  type: 'Identifier',
                  name: 'default',
                },
                computed: false,
                optional: false,
              },
              operator: '+',
              right: {
                type: 'Literal',
                value: hash,
                raw: `"${hash}"`,
              },
            },
          },
        ],
        sourceType: 'module',
        comments: [],
      },
    },
  };
}
