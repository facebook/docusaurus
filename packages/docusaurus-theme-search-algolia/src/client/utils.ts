/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import type {FacetFilters} from 'algoliasearch/lite';

export function mergeFacetFilters(
  f1: FacetFilters,
  f2: FacetFilters,
): FacetFilters;

export function mergeFacetFilters(
  f1: FacetFilters | undefined,
  f2: FacetFilters | undefined,
): FacetFilters | undefined;

export function mergeFacetFilters(
  f1: FacetFilters | undefined,
  f2: FacetFilters | undefined,
): FacetFilters | undefined {
  if (f1 === undefined) {
    return f2;
  }
  if (f2 === undefined) {
    return f1;
  }

  const normalize = (f: FacetFilters): FacetFilters =>
    typeof f === 'string' ? [f] : f;

  // Historical behavior: we flatten everything
  // TODO I'm pretty sure this is incorrect
  //  see https://www.algolia.com/doc/api-reference/api-parameters/facetFilters/?client=javascript
  //  Note: Algolia is working to provide a reliable facet merging strategy
  //  see https://github.com/facebook/docusaurus/pull/11327#issuecomment-3284742923
  return [...normalize(f1), ...normalize(f2)];
}
