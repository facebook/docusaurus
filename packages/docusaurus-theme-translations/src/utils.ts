/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

// This file isn't used by index.ts. It's used by update.mjs and tests. It's
// only here so that (a) we get a partially typed infrastructure (although the
// update script has ts-check anyways) (b) the test coverage isn't destroyed by
// the untested update.mjs file (c) we can ergonomically import the util
// functions in the Jest test without using `await import`

import path from 'path';
import fs from 'fs-extra';
// Unsafe import, should we create a package for the translationsExtractor ?;
import {
  globSourceCodeFilePaths,
  extractAllSourceCodeFileTranslations,
} from '@docusaurus/core/lib/server/translations/translationsExtractor';
import type {TranslationFileContent} from '@docusaurus/types';

async function getPackageCodePath(packageName: string) {
  const packagePath = path.join(__dirname, '../..', packageName);
  const packageJsonPath = path.join(packagePath, 'package.json');
  const {main} = (await fs.readJSON(packageJsonPath)) as {main: string};
  const packageSrcPath = path.join(packagePath, path.dirname(main));
  return packageSrcPath;
}

export async function getThemes(): Promise<{name: string; src: string[]}[]> {
  return [
    {
      name: 'theme-common',
      src: [
        await getPackageCodePath('docusaurus-theme-classic'),
        await getPackageCodePath('docusaurus-theme-common'),
      ],
    },
    {
      name: 'theme-search-algolia',
      src: [await getPackageCodePath('docusaurus-theme-search-algolia')],
    },
    {
      name: 'theme-live-codeblock',
      src: [await getPackageCodePath('docusaurus-theme-live-codeblock')],
    },
    {
      name: 'plugin-pwa',
      src: [await getPackageCodePath('docusaurus-plugin-pwa')],
    },
    {
      name: 'plugin-ideal-image',
      src: [await getPackageCodePath('docusaurus-plugin-ideal-image')],
    },
  ];
}

export async function extractThemeCodeMessages(
  targetDirs?: string[],
): Promise<TranslationFileContent> {
  // eslint-disable-next-line no-param-reassign
  targetDirs ??= (await getThemes()).flatMap((theme) => theme.src);

  const filePaths = (await globSourceCodeFilePaths(targetDirs)).filter(
    (filePath) => ['.js', '.jsx'].includes(path.extname(filePath)),
  );

  const filesExtractedTranslations = await extractAllSourceCodeFileTranslations(
    filePaths,
    {
      presets: [require.resolve('@docusaurus/core/lib/babel/preset')],
    },
  );

  filesExtractedTranslations.forEach((fileExtractedTranslations) => {
    if (fileExtractedTranslations.warnings.length > 0) {
      throw new Error(`
Please make sure all theme translations are static!
Some warnings were found!

${fileExtractedTranslations.warnings.join('\n\n')}
`);
    }
  });

  const translations = filesExtractedTranslations.reduce(
    (acc, extractedTranslations) => ({
      ...acc,
      ...extractedTranslations.translations,
    }),
    {},
  );

  return translations;
}
