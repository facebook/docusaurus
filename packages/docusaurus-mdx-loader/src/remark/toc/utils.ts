/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import {toValue} from '../utils';
import type {Node} from 'unist';
import type {
  MdxjsEsm,
  // @ts-expect-error: TODO see https://github.com/microsoft/TypeScript/issues/49721
} from 'mdast-util-mdx';
import type {TOCHeading, TOCItem, TOCItems, TOCSlice} from './types';
import type {
  Program,
  SpreadElement,
  ImportDeclaration,
  ImportSpecifier,
} from 'estree';

export function getImportDeclarations(program: Program): ImportDeclaration[] {
  return program.body.filter(
    (item): item is ImportDeclaration => item.type === 'ImportDeclaration',
  );
}

export function isMarkdownImport(node: Node): node is ImportDeclaration {
  if (node.type !== 'ImportDeclaration') {
    return false;
  }
  const importPath = (node as ImportDeclaration).source.value;
  return typeof importPath === 'string' && /\.mdx?$/.test(importPath);
}

export function findDefaultImportName(
  importDeclaration: ImportDeclaration,
): string | undefined {
  return importDeclaration.specifiers.find(
    (o: Node) => o.type === 'ImportDefaultSpecifier',
  )?.local.name;
}

export function findNamedImportSpecifier(
  importDeclaration: ImportDeclaration,
  localName: string,
): ImportSpecifier | undefined {
  return importDeclaration?.specifiers.find(
    (specifier): specifier is ImportSpecifier =>
      specifier.type === 'ImportSpecifier' &&
      specifier.local.name === localName,
  );
}

// Before: import Partial from "partial"
// After: import Partial, {toc as __tocPartial} from "partial"
export function addTocSliceImportIfNeeded({
  importDeclaration,
  tocExportName,
  tocSliceImportName,
}: {
  importDeclaration: ImportDeclaration;
  tocExportName: string;
  tocSliceImportName: string;
}): void {
  // We only add the toc slice named import if it doesn't exist already
  if (!findNamedImportSpecifier(importDeclaration, tocSliceImportName)) {
    importDeclaration.specifiers.push({
      type: 'ImportSpecifier',
      imported: {type: 'Identifier', name: tocExportName},
      local: {type: 'Identifier', name: tocSliceImportName},
    });
  }
}

export function isNamedExport(
  node: Node,
  exportName: string,
): node is MdxjsEsm {
  if (node.type !== 'mdxjsEsm') {
    return false;
  }
  const program = (node as MdxjsEsm).data?.estree;
  if (!program) {
    return false;
  }
  if (program.body.length !== 1) {
    return false;
  }
  const exportDeclaration = program.body[0]!;
  if (exportDeclaration.type !== 'ExportNamedDeclaration') {
    return false;
  }
  const variableDeclaration = exportDeclaration.declaration;
  if (variableDeclaration?.type !== 'VariableDeclaration') {
    return false;
  }
  const {id} = variableDeclaration.declarations[0]!;
  if (id.type !== 'Identifier') {
    return false;
  }

  return id.name === exportName;
}

export async function createTOCExportNodeAST({
  tocExportName,
  tocItems,
}: {
  tocExportName: string;
  tocItems: TOCItems;
}): Promise<MdxjsEsm> {
  function createTOCSliceAST(tocSlice: TOCSlice): SpreadElement {
    return {
      type: 'SpreadElement',
      argument: {type: 'Identifier', name: tocSlice.importName},
    };
  }

  async function createTOCHeadingAST({heading}: TOCHeading) {
    const {toString} = await import('mdast-util-to-string');
    const {valueToEstree} = await import('estree-util-value-to-estree');
    const value: TOCItem = {
      value: toValue(heading, toString),
      id: heading.data!.id!,
      level: heading.depth,
    };
    return valueToEstree(value);
  }

  async function createTOCItemAST(tocItem: TOCItems[number]) {
    switch (tocItem.type) {
      case 'slice':
        return createTOCSliceAST(tocItem);
      case 'heading':
        return createTOCHeadingAST(tocItem);
      default: {
        throw new Error(`unexpected toc item type`);
      }
    }
  }

  return {
    type: 'mdxjsEsm',
    value: '', // See https://github.com/facebook/docusaurus/pull/9684#discussion_r1457595181
    data: {
      estree: {
        type: 'Program',
        body: [
          {
            type: 'ExportNamedDeclaration',
            declaration: {
              type: 'VariableDeclaration',
              declarations: [
                {
                  type: 'VariableDeclarator',
                  id: {
                    type: 'Identifier',
                    name: tocExportName,
                  },
                  init: {
                    type: 'ArrayExpression',
                    elements: await Promise.all(tocItems.map(createTOCItemAST)),
                  },
                },
              ],
              kind: 'const',
            },
            specifiers: [],
            source: null,
          },
        ],
        sourceType: 'module',
      },
    },
  };
}
