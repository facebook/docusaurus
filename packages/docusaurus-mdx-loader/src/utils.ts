/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import fs from 'fs-extra';
import logger from '@docusaurus/logger';
import {escapePath, type WebpackCompilerName} from '@docusaurus/utils';
import {getProcessor, type SimpleProcessorResult} from './processor';
import {validateMDXFrontMatter} from './frontMatter';
import preprocessor from './preprocessor';
import type {Options} from './loader';

/**
 * When this throws, it generally means that there's no metadata file associated
 * with this MDX document. It can happen when using MDX partials (usually
 * starting with _). That's why it's important to provide the `isMDXPartial`
 * function in config
 */
export async function readMetadataPath(metadataPath: string): Promise<unknown> {
  try {
    return await fs.readJSON(metadataPath, 'utf8');
  } catch (error) {
    throw new Error(
      logger.interpolate`MDX loader can't read MDX metadata file path=${metadataPath}. Maybe the isMDXPartial option function was not provided?`,
      {cause: error as Error},
    );
  }
}

/**
 * Converts assets an object with Webpack require calls code.
 * This is useful for mdx files to reference co-located assets using relative
 * paths. Those assets should enter the Webpack assets pipeline and be hashed.
 * For now, we only handle that for images and paths starting with `./`:
 *
 * `{image: "./myImage.png"}` => `{image: require("./myImage.png")}`
 */
export function createAssetsExportCode({
  assets,
  inlineMarkdownAssetImageFileLoader,
}: {
  assets: unknown;
  inlineMarkdownAssetImageFileLoader: string;
}): string {
  if (
    typeof assets !== 'object' ||
    !assets ||
    Object.keys(assets).length === 0
  ) {
    return 'undefined';
  }

  // TODO implementation can be completed/enhanced
  function createAssetValueCode(assetValue: unknown): string | undefined {
    if (Array.isArray(assetValue)) {
      const arrayItemCodes = assetValue.map(
        (item: unknown) => createAssetValueCode(item) ?? 'undefined',
      );
      return `[${arrayItemCodes.join(', ')}]`;
    }
    // Only process string values starting with ./
    // We could enhance this logic and check if file exists on disc?
    if (typeof assetValue === 'string' && assetValue.startsWith('./')) {
      // TODO do we have other use-cases than image assets?
      // Probably not worth adding more support, as we want to move to Webpack 5 new asset system (https://github.com/facebook/docusaurus/pull/4708)
      return `require("${inlineMarkdownAssetImageFileLoader}${escapePath(
        assetValue,
      )}").default`;
    }
    return undefined;
  }

  const assetEntries = Object.entries(assets);

  const codeLines = assetEntries
    .map(([key, value]: [string, unknown]) => {
      const assetRequireCode = createAssetValueCode(value);
      return assetRequireCode ? `"${key}": ${assetRequireCode},` : undefined;
    })
    .filter(Boolean);

  return `{\n${codeLines.join('\n')}\n}`;
}

/**
 * data.contentTitle is set by the remark contentTitle plugin
 */
export function extractContentTitleData(data: {
  [key: string]: unknown;
}): string | undefined {
  return data.contentTitle as string | undefined;
}

export async function compileToJSX({
  filePath,
  fileContent,
  frontMatter,
  options,
  compilerName,
}: {
  filePath: string;
  fileContent: string;
  frontMatter: Record<string, unknown>;
  options: Options;
  compilerName: WebpackCompilerName;
}): Promise<SimpleProcessorResult> {
  const preprocessedFileContent = preprocessor({
    fileContent,
    filePath,
    admonitions: options.admonitions,
    markdownConfig: options.markdownConfig,
  });

  const mdxFrontMatter = validateMDXFrontMatter(frontMatter.mdx);

  const processor = await getProcessor({
    filePath,
    options,
    mdxFrontMatter,
  });

  try {
    return await processor.process({
      content: preprocessedFileContent,
      filePath,
      frontMatter,
      compilerName,
    });
  } catch (errorUnknown) {
    const error = errorUnknown as Error;

    // MDX can emit errors that have useful extra attributes
    const errorJSON = JSON.stringify(error, null, 2);
    const errorDetails =
      errorJSON === '{}'
        ? // regular JS error case: print stacktrace
          error.stack ?? 'N/A'
        : // MDX error: print extra attributes + stacktrace
          `${errorJSON}\n${error.stack}`;

    throw new Error(
      `MDX compilation failed for file ${logger.path(filePath)}\nCause: ${
        error.message
      }\nDetails:\n${errorDetails}`,
      // TODO error cause doesn't seem to be used by Webpack stats.errors :s
      {cause: error},
    );
  }
}
