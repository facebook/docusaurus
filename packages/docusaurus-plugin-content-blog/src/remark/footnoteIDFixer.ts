/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import {simpleHash} from '@docusaurus/utils';
// @ts-expect-error: TODO see https://github.com/microsoft/TypeScript/issues/49721
import type {Transformer} from 'unified';
import type {FootnoteReference, FootnoteDefinition} from 'mdast';

/**
 * In the blog list view, each post will be compiled separately. However, they
 * may use the same footnote IDs. This leads to duplicated DOM IDs and inability
 * to navigate to footnote references. This plugin fixes it by appending a
 * unique hash to each reference/definition.
 */
export default function plugin(): Transformer {
  return async (root, vfile) => {
    const {visit} = await import('unist-util-visit');

    const suffix = `-${simpleHash(vfile.path!, 6)}`;
    visit(root, 'footnoteReference', (node: FootnoteReference) => {
      node.identifier += suffix;
    });
    visit(root, 'footnoteDefinition', (node: FootnoteDefinition) => {
      node.identifier += suffix;
    });
  };
}
