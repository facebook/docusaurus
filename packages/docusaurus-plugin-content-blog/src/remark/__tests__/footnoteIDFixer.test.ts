/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 */

import path from 'path';
import vfile from 'to-vfile';

import {simpleHash} from '@docusaurus/utils';
import footnoteIDFixer from '../footnoteIDFixer';

const processFixture = async (name: string) => {
  const mdx = await import('@mdx-js/mdx');
  const {default: gfm} = await import('remark-gfm');

  const filePath = path.join(__dirname, `__fixtures__/${name}.md`);
  const file = await vfile.read(filePath);

  const result = await mdx.compile(file, {
    remarkPlugins: [gfm, footnoteIDFixer],
  });

  return result.value;
};

describe('footnoteIDFixer remark plugin', () => {
  it('appends a hash to each footnote def/ref', async () => {
    const hash = simpleHash(path.join(__dirname, `__fixtures__/post.md`), 6);
    expect(
      (await processFixture('post')).replace(new RegExp(hash, 'g'), '[HASH]'),
    ).toMatchSnapshot();
  });

  it('produces different hashes for different posts but same hash for the same path', async () => {
    const file1 = await processFixture('post');
    const file1again = await processFixture('post');
    const file2 = await processFixture('post2');
    expect(file1).toBe(file1again);
    expect(file1).not.toBe(file2);
  });
});
