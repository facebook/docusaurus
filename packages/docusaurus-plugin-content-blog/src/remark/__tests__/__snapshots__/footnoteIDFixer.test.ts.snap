// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`footnoteIDFixer remark plugin appends a hash to each footnote def/ref 1`] = `
"import {Fragment as _Fragment, jsx as _jsx, jsxs as _jsxs} from "react/jsx-runtime";
function _createMdxContent(props) {
  const _components = {
    a: "a",
    h2: "h2",
    li: "li",
    ol: "ol",
    p: "p",
    section: "section",
    sup: "sup",
    ...props.components
  };
  return _jsxs(_Fragment, {
    children: [_jsxs(_components.p, {
      children: ["foo", _jsx(_components.sup, {
        children: _jsx(_components.a, {
          href: "#user-content-fn-1-[HASH]",
          id: "user-content-fnref-1-[HASH]",
          "data-footnote-ref": true,
          "aria-describedby": "footnote-label",
          children: "1"
        })
      })]
    }), "/n", _jsxs(_components.p, {
      children: ["bar", _jsx(_components.sup, {
        children: _jsx(_components.a, {
          href: "#user-content-fn-2-[HASH]",
          id: "user-content-fnref-2-[HASH]",
          "data-footnote-ref": true,
          "aria-describedby": "footnote-label",
          children: "2"
        })
      })]
    }), "/n", _jsxs(_components.p, {
      children: ["baz", _jsx(_components.sup, {
        children: _jsx(_components.a, {
          href: "#user-content-fn-3-[HASH]",
          id: "user-content-fnref-3-[HASH]",
          "data-footnote-ref": true,
          "aria-describedby": "footnote-label",
          children: "3"
        })
      })]
    }), "/n", _jsxs(_components.section, {
      "data-footnotes": true,
      className: "footnotes",
      children: [_jsx(_components.h2, {
        className: "sr-only",
        id: "footnote-label",
        children: "Footnotes"
      }), "/n", _jsxs(_components.ol, {
        children: ["/n", _jsxs(_components.li, {
          id: "user-content-fn-1-[HASH]",
          children: ["/n", _jsxs(_components.p, {
            children: ["foo ", _jsx(_components.a, {
              href: "#user-content-fnref-1-[HASH]",
              "data-footnote-backref": "",
              "aria-label": "Back to reference 1",
              className: "data-footnote-backref",
              children: "↩"
            })]
          }), "/n"]
        }), "/n", _jsxs(_components.li, {
          id: "user-content-fn-2-[HASH]",
          children: ["/n", _jsxs(_components.p, {
            children: ["foo ", _jsx(_components.a, {
              href: "#user-content-fnref-2-[HASH]",
              "data-footnote-backref": "",
              "aria-label": "Back to reference 2",
              className: "data-footnote-backref",
              children: "↩"
            })]
          }), "/n"]
        }), "/n", _jsxs(_components.li, {
          id: "user-content-fn-3-[HASH]",
          children: ["/n", _jsxs(_components.p, {
            children: ["foo ", _jsx(_components.a, {
              href: "#user-content-fnref-3-[HASH]",
              "data-footnote-backref": "",
              "aria-label": "Back to reference 3",
              className: "data-footnote-backref",
              children: "↩"
            })]
          }), "/n"]
        }), "/n"]
      }), "/n"]
    })]
  });
}
export default function MDXContent(props = {}) {
  const {wrapper: MDXLayout} = props.components || ({});
  return MDXLayout ? _jsx(MDXLayout, {
    ...props,
    children: _jsx(_createMdxContent, {
      ...props
    })
  }) : _createMdxContent(props);
}
"
`;
